src/components/FlatTreeRenderer.js

```diff
@@ -0,0 +1,114 @@
+  }, 0) - 1
+
+  const oldInvisibleThoughtsAboveCountRef = React.useRef(invisibleThoughtsAboveCount)
+
+  React.useEffect(() => {
+    // storing prev invisible thought counts to compare between re-render
+    oldInvisibleThoughtsAboveCountRef.current = invisibleThoughtsAboveCount
+  })
+
+  return (
+    <div style={{ position: 'relative', marginLeft: '10rem', marginTop: '5rem' }}>
+      <TransitionGroup>
+        {flatArray.map((node, i) => {
+
+          const style = { transition: `opacity 750ms ease-in-out, transform 750ms ease-in-out`, position: 'absolute', top: `${(invisibleThoughtsAboveCount + i) * 30}px`, display: 'block', margin: '0.5rem 0' }
+
+          /*
+            unique key for each nodes.
+            current implementation doesn't account for two thoughts with same value within same context
+           */
+          const key = `${contextOf(node.path).value}-${node.value}-${node.path.length}`
+
+          const nodeLeft = `${(node.path.length - visibleDepth) * 1.2}rem`
+          const parentLeft = `${Math.max((node.path.length - visibleDepth + 1) * 1.2, 0)}rem`
+
+          /*
+            some children nodes on mount should animate their left position based on the direct parent current animation state
+
+            For example:
+
+            - a
+              - b
+              - c
+              - e
+
+            on selecting thought c
+
+            -a
+              - b
+              - c
+                - d
+              - e
+
+            At such cases where c will start animating to the left, and its children thought d on mount should start animating along with parent c.
+
+            With current implementation animation only works properly if parent doesn't have on going animation. If parent is already animating
+            then we don't have access to animation state of the parent to adjust childrens animation.
+          */
+          const animateIncomingNodeAccToParent = node.isCursorChildren && invisibleThoughtsAboveCount !== oldInvisibleThoughtsAboveCountRef.current
+
+          const transitionStyles = {
+            entering: { opacity: node.isCursorChildren ? 1 : 0, transform: `translateX(${animateIncomingNodeAccToParent ? parentLeft : nodeLeft})` },
+            entered: { opacity: node.isDistantThought ? 0.35 : 1, transform: `translateX(${nodeLeft})` },
+            exiting: { opacity: 0, transform: `translateX(${nodeLeft})` },
+          }
+
+          // dynamically changing timeout to stop exit animation for some cases
+          // still lacks proper logic for handling the exit animation
+          return (
+            <Transition key={key} timeout={{ enter: 0, exit: node.isCursorChildren ? 0 : 750 }} onExit={el => {
+              // we have to forcibly hide thoughts that are exiting due to an ancestor collapse so that the default animation is skipped
+              if (!node.isDistantThought) {
+                el.style.display = 'none'
+              }
+            }}>
```

Comment:
@shresthabijay This will allow certain thoughts to skip the exit animation. I don't know if this is the right condition, but the technique works.

Reply:
@raineorshine I will get back to you after I try this. Thank you!